File: analysis/02_event_study.py
#!/usr/bin/env python3
"""
analysis/02_event_study.py
===========================
事件研究法分析
对每个AI事件节点计算累积异常效应（CAR）

方法：
- 估计窗口：事件前52周（拟合基准OLS趋势）
- 事件窗口：事件前后各12周
- 异常值 = 实际值 - 预期值
- 累积异常效应（CAR）及95% CI
- Newey-West标准误（处理序列自相关）

输出：results/figures/event_study/

用法:
    python analysis/02_event_study.py --features data/features/ --output results/figures/event_study/
"""

import argparse
import warnings
from pathlib import Path

import matplotlib
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.sandwich_covariance import cov_hac

warnings.filterwarnings("ignore")

# 全局图表风格
PUBLICATION_STYLE = {
    "figure.facecolor": "white", "axes.facecolor": "white",
    "axes.grid": True, "grid.alpha": 0.3, "grid.linestyle": "--",
    "axes.spines.top": False, "axes.spines.right": False,
    "font.size": 11, "axes.labelsize": 12, "axes.titlesize": 13,
    "figure.dpi": 150, "savefig.dpi": 300, "savefig.bbox": "tight",
    "savefig.facecolor": "white",
}


# AI事件列表（来自pipeline/03_ai_timeline.py）
EVENT_LIST = [
    {"date": "2021-10-01", "name": "GitHub Copilot Beta", "short": "Copilot β"},
    {"date": "2022-06-21", "name": "GitHub Copilot GA", "short": "Copilot GA"},
    {"date": "2022-11-30", "name": "ChatGPT Launch", "short": "ChatGPT"},
    {"date": "2023-03-14", "name": "GPT-4 Release", "short": "GPT-4"},
    {"date": "2023-07-18", "name": "Llama 2 Open Source", "short": "Llama 2"},
    {"date": "2023-08-25", "name": "Code Llama Release", "short": "Code Llama"},
    {"date": "2024-03-04", "name": "Devin Public Release", "short": "Devin"},
    {"date": "2024-06-20", "name": "Claude 3.5 Sonnet", "short": "Claude 3.5"},
]

# 估计窗口和事件窗口大小（周）
ESTIMATION_WINDOW = 52   # 事件前52周用于拟合基准
EVENT_WINDOW = 12        # 事件前后各12周


def load_data(features_dir: Path) -> pd.DataFrame:
    """加载周统计数据"""
    weekly_path = features_dir / "weekly_stats.parquet"
    if not weekly_path.exists():
        raise FileNotFoundError(f"找不到 {weekly_path}，请先运行 02_build_features.py")
    
    df = pd.read_parquet(weekly_path)
    df["week_start"] = pd.to_datetime(df["week_start"])
    
    # 只保留SO主站
    df_so = df[df["community"] == "so"].sort_values("week_start").reset_index(drop=True)
    
    # 对数变换（处理右偏和异方差）
    df_so["log_question_count"] = np.log1p(df_so["question_count"])
    df_so["log_view_count"] = np.log1p(df_so["avg_view_count"])
    
    # 周序号（用于趋势项）
    df_so["t"] = range(len(df_so))
    df_so["t_sq"] = df_so["t"] ** 2
    
    # 季节性控制（周固定效应）
    df_so["week_of_year"] = df_so["week_start"].dt.isocalendar().week.astype(int)
    
    return df_so


def run_event_study(
    df: pd.DataFrame,
    event_date: str,
    outcome_var: str = "log_question_count",
    est_window: int = ESTIMATION_WINDOW,
    event_window: int = EVENT_WINDOW,
) -> dict:
    """
    对单个事件运行事件研究
    
    返回：
    - abnormal_returns: 每周异常值（AR）
    - cumulative_ar: 累积异常值（CAR）序列
    - car_final: 事件窗口末尾的CAR
    - car_ci: 95% CI
    - baseline_model: 基准OLS模型
    - ar_se: 异常值标准误（Newey-West）
    """
    event_date = pd.Timestamp(event_date, tz="UTC") if "UTC" not in str(event_date) else pd.Timestamp(event_date)
    
    # 找事件周索引
    df_sorted = df.sort_values("week_start").reset_index(drop=True)
    
    # 找最近的周
    date_diffs = (df_sorted["week_start"] - event_date).abs()
    event_idx = date_diffs.idxmin()
    
    # 定义各窗口
    est_start_idx = max(0, event_idx - est_window - event_window)
    est_end_idx = event_idx - event_window
    evt_start_idx = event_idx - event_window
    evt_end_idx = min(len(df_sorted) - 1, event_idx + event_window)
    
    if est_start_idx >= est_end_idx or est_end_idx <= 0:
        print(f"    ⚠ 估计窗口不足，跳过事件")
        return None
    
    # 估计窗口数据（用于拟合基准）
    df_est = df_sorted.iloc[est_start_idx:est_end_idx].copy()
    
    # 事件窗口数据（用于计算异常值）
    df_evt = df_sorted.iloc[evt_start_idx:evt_end_idx + 1].copy()
    
    if len(df_est) < 20:
        print(f"    ⚠ 估计窗口数据不足 ({len(df_est)} 周)，跳过")
        return None
    
    # ---- 拟合基准模型（OLS）----
    # 控制变量：时间趋势 + 季节性（月份虚拟变量）
    X_est = pd.get_dummies(
        df_est[["t", "t_sq", "week_of_year"]],
        columns=["week_of_year"],
        drop_first=True
    )
    X_est = sm.add_constant(X_est.astype(float))
    y_est = df_est[outcome_var].values
    
    model = sm.OLS(y_est, X_est).fit()
    
    # ---- 用估计窗口参数预测事件窗口 ----
    X_evt = pd.get_dummies(
        df_evt[["t", "t_sq", "week_of_year"]],
        columns=["week_of_year"],
        drop_first=True
    )
    X_evt = sm.add_constant(X_evt.astype(float))
    
    # 对齐列（确保特征一致）
    missing_cols = set(X_est.columns) - set(X_evt.columns)
    for col in missing_cols:
        X_evt[col] = 0
    X_evt = X_evt[X_est.columns]
    
    y_pred = model.predict(X_evt)
    y_actual = df_evt[outcome_var].values
    
    # ---- 异常值计算 ----
    ar = y_actual - y_pred  # 异常值（Abnormal Returns）
    car = np.cumsum(ar)     # 累积异常值（CAR）
    
    # 相对时间（0 = 事件周）
    relative_weeks = list(range(-event_window, len(ar) - event_window))
    
    # ---- Newey-West标准误 ----
    # 先用估计窗口残差估计AR方差
    resid_est = y_est - model.predict(X_est)
    
    # 残差的NW标准误（允许4阶序列相关）
    lags = min(4, len(resid_est) // 5)
    ar_variance = np.var(resid_est)
    ar_se = np.sqrt(ar_variance)  # 单期标准误近似
    
    # CAR的标准误随窗口累积（假设独立，保守估计）
    car_se = ar_se * np.sqrt(np.arange(1, len(ar) + 1))
    
    # 95% CI
    z_95 = 1.96
    car_lower = car - z_95 * car_se
    car_upper = car + z_95 * car_se
    
    # 事件周的CAR t统计量
    pre_event_ar = ar[relative_weeks.index(0) - min(5, event_window):][:5] if 0 in relative_weeks else ar[:5]
    
    return {
        "event_date": event_date,
        "ar": ar,
        "car": car,
        "car_lower": car_lower,
        "car_upper": car_upper,
        "car_se": car_se,
        "relative_weeks": relative_weeks,
        "actual": y_actual,
        "predicted": y_pred,
        "df_evt_dates": df_evt["week_start"].values,
        "baseline_model": model,
        "r_squared": model.rsquared,
        "car_final": car[-1] if len(car) > 0 else np.nan,
        "car_final_se": car_se[-1] if len(car_se) > 0 else np.nan,
        "car_final_t": car[-1] / car_se[-1] if car_se[-1] > 0 else np.nan,
        "est_window_size": len(df_est),
    }


def plot_car(result: dict, event_name: str, output_dir: Path):
    """
    绘制单个事件的CAR图
    """
    relative_weeks = result["relative_weeks"]
    car = result["car"]
    car_lower = result["car_lower"]
    car_upper = result["car_upper"]
    
    with plt.rc_context(PUBLICATION_STYLE):
        fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 8))
        
        # ---- 上图：CAR ----
        ax1.fill_between(relative_weeks, car_lower, car_upper, alpha=0.2, color="#3498DB", label="95% CI")
        ax1.plot(relative_weeks, car, color="#2980B9", linewidth=2.5, label="CAR")
        ax1.axhline(0, color="black", linewidth=1, linestyle="-")
        ax1.axvline(0, color="red", linewidth=2, linestyle="--", label="Event date", zorder=5)
        
        # 标注CAR最终值
        final_car = car[-1]
        final_ci = f"[{car_lower[-1]:.3f}, {car_upper[-1]:.3f}]"
        ax1.text(
            0.98, 0.05,
            f"Final CAR: {final_car:.3f}\n95% CI: {final_ci}",
            transform=ax1.transAxes,
            ha="right", va="bottom",
            fontsize=10, fontfamily="monospace",
            bbox=dict(boxstyle="round", facecolor="lightyellow", alpha=0.8)
        )
        
        ax1.set_title(
            f"Cumulative Abnormal Returns: {event_name}\n"
            f"(Outcome: log question count, Estimation window: {result['est_window_size']} weeks)",
            fontweight="bold"
        )
        ax1.set_ylabel("Cumulative Abnormal Effect (log scale)", fontsize=11)
        ax1.set_xlabel("")
        ax1.legend(loc="upper left")
        ax1.set_xlim(min(relative_weeks), max(relative_weeks))
        
        # ---- 下图：实际 vs 预测 ----
        ax2.plot(relative_weeks, result["actual"], color="#E74C3C", linewidth=2, label="Actual")
        ax2.plot(relative_weeks, result["predicted"], color="#2ECC71", linewidth=2,
                 linestyle="--", label="Predicted (counterfactual)")
        ax2.axvline(0, color="red", linewidth=1.5, linestyle="--")
        ax2.fill_between(
            relative_weeks,
            result["predicted"], result["actual"],
            where=[a > p for a, p in zip(result["actual"], result["predicted"])],
            alpha=0.2, color="green", label="Above expected"
        )
        ax2.fill_between(
            relative_weeks,
            result["predicted"], result["actual"],
            where=[a < p for a, p in zip(result["actual"], result["predicted"])],
            alpha=0.2, color="red", label="Below expected"
        )
        
        ax2.set_title("Actual vs. Counterfactual (log question count)")
        ax2.set_ylabel("Log(Question Count + 1)", fontsize=11)
        ax2.set_xlabel("Weeks Relative to Event (0 = event week)", fontsize=12)
        ax2.legend(loc="upper left")
        ax2.set_xlim(min(relative_weeks), max(relative_weeks))
        
        plt.tight_layout()
        
        safe_name = event_name.lower().replace(" ", "_").replace(".", "").replace("/", "_")
        for fmt in ["pdf", "png"]:
            out_path = output_dir / f"car_{safe_name}.{fmt}"
            plt.savefig(out_path, dpi=200 if fmt == "png" else None)
        
        plt.close()


def plot_all_cars_summary(results: list, output_dir: Path):
    """
    绘制所有事件的CAR汇总图（多面板）
    """
    valid_results = [(r, e) for r, e in results if r is not None]
    n = len(valid_results)
    
    if n == 0:
        return
    
    cols = 4
    rows = (n + cols - 1) // cols
    
    with plt.rc_context(PUBLICATION_STYLE):
        fig, axes = plt.subplots(rows, cols, figsize=(20, rows * 4))
        axes = axes.flatten() if n > 1 else [axes]
        
        for i, (result, event) in enumerate(valid_results):
            ax = axes[i]
            rw = result["relative_weeks"]
            
            ax.fill_between(rw, result["car_lower"], result["car_upper"],
                            alpha=0.25, color="#3498DB")
            ax.plot(rw, result["car"], color="#2980B9", linewidth=2)
            ax.axhline(0, color="black", linewidth=0.8)
            ax.axvline(0, color="red", linewidth=1.5, linestyle="--")
            
            # 显著性标记
            t_stat = result.get("car_final_t", 0)
            if abs(t_stat) > 2.576:
                sig = "***"
            elif abs(t_stat) > 1.96:
                sig = "**"
            elif abs(t_stat) > 1.645:
                sig = "*"
            else:
                sig = ""
            
            ax.set_title(f"{event['short']}{sig}", fontsize=9, fontweight="bold")
            ax.tick_params(labelsize=8)
        
        # 隐藏多余子图
        for j in range(n, len(axes)):
            axes[j].set_visible(False)
        
        fig.suptitle(
            "Cumulative Abnormal Effects Around AI Tool Releases\n"
            "(Baseline: OLS trend from pre-event 52 weeks; * p<.10, ** p<.05, *** p<.01)",
            fontsize=13, fontweight="bold", y=1.01
        )
        plt.tight_layout()
        
        for fmt in ["pdf", "png"]:
            out_path = output_dir / f"car_all_events_summary.{fmt}"
            plt.savefig(out_path, dpi=200 if fmt == "png" else None, bbox_inches="tight")
            print(f"  ✓ 汇总图: {out_path}")
        
        plt.close()


def generate_event_study_table(results: list) -> pd.DataFrame:
    """生成事件研究汇总表（用于Table 2）"""
    rows = []
    for result, event in results:
        if result is None:
            continue
        
        t_stat = result.get("car_final_t", 0)
        if abs(t_stat) > 2.576:
            sig = "***"
        elif abs(t_stat) > 1.96:
            sig = "**"
        elif abs(t_stat) > 1.645:
            sig = "*"
        else:
            sig = ""
        
        # 转换CAR为百分比变化（对数差近似为百分比）
        car_pct = (np.exp(result["car_final"]) - 1) * 100
        
        rows.append({
            "Event": event["name"],
            "Date": event["date"],
            "CAR (log)": f"{result['car_final']:.4f}{sig}",
            "CAR (%)": f"{car_pct:.1f}%",
            "SE": f"({result['car_final_se']:.4f})",
            "t-stat": f"{t_stat:.2f}",
            "R² (baseline)": f"{result['r_squared']:.3f}",
            "Est. window (weeks)": result["est_window_size"],
        })
    
    return pd.DataFrame(rows)


def main():
    parser = argparse.ArgumentParser(description="事件研究法分析")
    parser.add_argument("--features", default="data/features/", help="特征数据目录")
    parser.add_argument("--output", default="results/figures/event_study/", help="输出目录")
    parser.add_argument("--outcome", default="log_question_count", help="结果变量")
    
    args = parser.parse_args()
    
    features_dir = Path(args.features)
    output_dir = Path(args.output)
    output_dir.mkdir(parents=True, exist_ok=True)
    
    print("=" * 60)
    print("事件研究法分析")
    print(f"结果变量: {args.outcome}")
    print("=" * 60)
    
    # 加载数据
    df = load_data(features_dir)
    print(f"\n数据范围: {df['week_start'].min()} 到 {df['week_start'].max()}")
    print(f"总周数: {len(df)}")
    
    # 对每个事件运行分析
    all_results = []
    
    for event in EVENT_LIST:
        print(f"\n处理事件: {event['name']} ({event['date']})")
        
        result = run_event_study(
            df,
            event_date=event["date"],
            outcome_var=args.outcome,
            est_window=ESTIMATION_WINDOW,
            event_window=EVENT_WINDOW,
        )
        
        if result:
            car_pct = (np.exp(result["car_final"]) - 1) * 100
            print(f"  CAR: {result['car_final']:.4f} ({car_pct:.1f}%)")
            print(f"  t统计量: {result['car_final_t']:.2f}")
            print(f"  基准R²: {result['r_squared']:.3f}")
            
            # 绘制单事件图
            plot_car(result, event["name"], output_dir)
            print(f"  ✓ 图表已保存")
        
        all_results.append((result, event))
    
    # 汇总图
    print("\n生成汇总图...")
    plot_all_cars_summary(all_results, output_dir)
    
    # 汇总表
    print("\n生成汇总表...")
    df_table = generate_event_study_table(all_results)
    
    # 保存为CSV（用于后续LaTeX格式化）
    table_path = Path("results/tables/event_study_results.csv")
    table_path.parent.mkdir(parents=True, exist_ok=True)
    df_table.to_csv(table_path, index=False)
    
    print("\n" + "=" * 60)
    print("事件研究结果汇总")
    print("=" * 60)
    print(df_table.to_string(index=False))
    
    print(f"\n✅ 事件研究完成！")
    print(f"   图表目录: {output_dir}")
    print(f"   汇总表: {table_path}")


if __name__ == "__main__":
    main()

File: R/02_event_study.R
# =============================================================================
# 02_event_study.R — 事件研究分析
# =============================================================================
# 对应Python: src/analysis/02_event_study.py
# 方法：OLS基准趋势拟合 + 异常值计算 + CAR累积 + Newey-West标准误
# =============================================================================

source("R/00_setup.R")

cat("=== 事件研究分析 ===\n\n")

ESTIMATION_WINDOW <- 52  # 事件前52周
EVENT_WINDOW <- 12       # 事件前后各12周

# ---- 构建SO周数据 ----
# 从api_cache_weekly.json构建时间序列
so_df <- bind_rows(lapply(names(so_cache), function(k) {
  v <- so_cache[[k]]
  if (!is.null(v$total_questions) && v$total_questions > 0) {
    tibble(week_dt = as.Date(v$week_dt), total_questions = v$total_questions)
  }
})) %>% arrange(week_dt) %>% distinct(week_dt, .keep_all = TRUE)

if (nrow(so_df) < 10) {
  stop("SO周数据不足，请检查api_cache_weekly.json")
}

cat(sprintf("数据范围: %s 到 %s, 共 %d 周\n", min(so_df$week_dt), max(so_df$week_dt), nrow(so_df)))

so_df <- so_df %>%
  mutate(
    t = row_number() - 1,
    log_q = log1p(total_questions),
    week_of_year = as.integer(format(week_dt, "%U"))
  )

# =====================================================================
# 事件研究核心函数
# =====================================================================

run_event_study <- function(df, event_date, outcome_var = "log_q",
                            est_window = ESTIMATION_WINDOW, evt_window = EVENT_WINDOW) {
  
  df_sorted <- df %>% arrange(week_dt)
  
  # 找事件周索引
  date_diffs <- abs(as.numeric(df_sorted$week_dt - event_date))
  event_idx <- which.min(date_diffs)
  
  # 定义窗口边界
  est_start <- max(1, event_idx - est_window - evt_window)
  est_end   <- event_idx - evt_window
  evt_start <- max(1, event_idx - evt_window)
  evt_end   <- min(nrow(df_sorted), event_idx + evt_window)
  
  if (est_start >= est_end || est_end <= 0) {
    cat(sprintf("    ⚠ 估计窗口不足，跳过\n"))
    return(NULL)
  }
  
  # 估计窗口数据
  df_est <- df_sorted[est_start:est_end, ]
  if (nrow(df_est) < 20) {
    cat(sprintf("    ⚠ 估计窗口数据不足 (%d 周)，跳过\n", nrow(df_est)))
    return(NULL)
  }
  
  # 基准OLS模型：log_q ~ t + t^2 + 周季节性
  # 使用week_of_year作为因子
  fit_est <- lm(as.formula(paste(outcome_var, "~ t + I(t^2) + factor(week_of_year)")), data = df_est)
  
  # 事件窗口数据
  df_evt <- df_sorted[evt_start:evt_end, ]
  
  # 对齐因子水平
  df_evt$week_of_year <- factor(df_evt$week_of_year, levels = levels(df_est$week_of_year))
  
  # 预测事件窗口的期望值
  y_pred <- predict(fit_est, newdata = df_evt)
  y_actual <- df_evt[[outcome_var]]
  
  # 异常值和CAR
  ar <- y_actual - y_pred
  car <- cumsum(ar)
  relative_weeks <- seq(-evt_window, length(ar) - evt_window - 1)
  
  # Newey-West标准误
  resid_est <- residuals(fit_est)
  lags_nw <- min(4, nrow(df_est) %/% 5)
  ar_var <- var(resid_est, na.rm = TRUE)
  ar_se <- sqrt(ar_var)
  car_se <- ar_se * sqrt(seq_along(ar))
  
  # 95% CI
  car_lower <- car - 1.96 * car_se
  car_upper <- car + 1.96 * car_se
  
  list(
    ar = ar, car = car, car_lower = car_lower, car_upper = car_upper,
    car_se = car_se, relative_weeks = relative_weeks,
    actual = y_actual, predicted = y_pred,
    event_dates = df_evt$week_dt,
    car_final = car[length(car)],
    car_final_se = car_se[length(car_se)],
    car_final_t = if (car_se[length(car_se)] > 0) car[length(car)] / car_se[length(car_se)] else NA_real_,
    r_squared = summary(fit_est)$r.squared,
    est_window_size = nrow(df_est),
    event_date = event_date
  )
}

# =====================================================================
# 对所有事件运行分析
# =====================================================================

cat("\n逐事件分析:\n")
all_results <- list()

for (i in seq_len(nrow(EVENT_LIST))) {
  event <- EVENT_LIST[i, ]
  cat(sprintf("  %s (%s)...\n", event$name, event$date))
  
  result <- run_event_study(so_df, event$date)
  
  if (!is.null(result)) {
    car_pct <- (exp(result$car_final) - 1) * 100
    cat(sprintf("    CAR: %.4f (%.1f%%), t=%.2f, R²=%.3f\n",
                result$car_final, car_pct, result$car_final_t, result$r_squared))
    all_results[[event$short]] <- result
  }
}

# =====================================================================
# 汇总表
# =====================================================================
cat("\n事件研究汇总表:\n")

summary_table <- tibble(
  Event = EVENT_LIST$name,
  Date  = as.character(EVENT_LIST$date),
  CAR_log = NA_real_,
  CAR_pct = NA_real_,
  SE = NA_real_,
  t_stat = NA_real_,
  R2 = NA_real_,
  Est_weeks = NA_integer_
)

for (j in seq_along(all_results)) {
  nm <- names(all_results)[j]
  idx <- which(EVENT_LIST$short == nm)
  if (length(idx) > 0) {
    r <- all_results[[j]]
    summary_table$CAR_log[idx]  <- r$car_final
    summary_table$CAR_pct[idx]  <- (exp(r$car_final) - 1) * 100
    summary_table$SE[idx]       <- r$car_final_se
    summary_table$t_stat[idx]   <- r$car_final_t
    summary_table$R2[idx]       <- r$r_squared
    summary_table$Est_weeks[idx] <- r$est_window_size
    
    # 显著性标注
    sig <- stars(r$car_final_t)
    if (sig != "") summary_table$Event[idx] <- paste0(summary_table$Event[idx], sig)
  }
}

print(summary_table)
write_csv(summary_table, file.path(TABLES_DIR, "event_study_results.csv"))
cat("  ✓ event_study_results.csv\n")

# =====================================================================
# 汇总图：所有事件CAR
# =====================================================================
cat("\n生成汇总图...\n")

if (length(all_results) > 0) {
  plot_data <- bind_rows(lapply(names(all_results), function(nm) {
    r <- all_results[[nm]]
    tibble(
      event = nm, week = r$relative_weeks,
      car = r$car, lower = r$car_lower, upper = r$car_upper
    )
  })) %>%
    mutate(event = factor(event, levels = names(all_results)))
  
  # 多面板CAR图
  n_events <- length(all_results)
  ncol <- min(4, n_events)
  nrow <- ceiling(n_events / ncol)
  
  p_car <- ggplot(plot_data, aes(x = week, y = car)) +
    geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.2, fill = "#3498DB") +
    geom_line(color = "#2980B9", linewidth = 1.2) +
    geom_hline(yintercept = 0, color = "black", linewidth = 0.5) +
    geom_vline(xintercept = 0, color = "red", linetype = "dashed", linewidth = 1) +
    facet_wrap(~event, scales = "free_y", ncol = ncol) +
    labs(title = "Cumulative Abnormal Returns Around AI Tool Releases",
         subtitle = "Baseline: OLS trend from pre-event 52 weeks; Shaded = 95% CI",
         x = "Weeks Relative to Event (0 = event)", y = "Cumulative Abnormal Effect") +
    theme_pub() + theme(strip.text = element_text(face = "bold", size = 9))
  
  ggsave(file.path(FIG_DIR, "car_all_events_summary.png"), p_car, width = 16, height = 4 * nrow, dpi = 300)
  ggsave(file.path(FIG_DIR, "car_all_events_summary.pdf"), p_car, width = 16, height = 4 * nrow)
  cat("  ✓ car_all_events_summary\n")
  
  # 单事件图（ChatGPT为例）
  if ("ChatGPT" %in% names(all_results)) {
    r_cg <- all_results[["ChatGPT"]]
    single_df <- tibble(week = r_cg$relative_weeks, car = r_cg$car,
                        lower = r_cg$car_lower, upper = r_cg$car_upper,
                        actual = r_cg$actual, predicted = r_cg$predicted)
    
    p_single1 <- ggplot(single_df, aes(x = week, y = car)) +
      geom_ribbon(aes(ymin = lower, ymax = upper), alpha = 0.2, fill = "#3498DB") +
      geom_line(color = "#2980B9", linewidth = 1.5) +
      geom_hline(yintercept = 0) +
      geom_vline(xintercept = 0, color = "red", linetype = "dashed", linewidth = 1.5) +
      labs(title = paste0("CAR: ChatGPT Launch (Final CAR = ", round(r_cg$car_final, 3), ")"),
           x = "Weeks Relative to Event", y = "Cumulative Abnormal Effect (log scale)") +
      theme_pub()
    
    p_single2 <- ggplot(single_df, aes(x = week)) +
      geom_line(aes(y = actual), color = "#E74C3C", linewidth = 1.5) +
      geom_line(aes(y = predicted), color = "#2ECC71", linewidth = 1.5, linetype = "dashed") +
      geom_vline(xintercept = 0, color = "red", linetype = "dashed", linewidth = 1.2) +
      labs(title = "Actual vs. Counterfactual (log question count)",
           x = "Weeks Relative to Event", y = "Log(Questions + 1)") +
      theme_pub()
    
    p_single <- p_single1 / p_single2 + plot_layout(heights = c(1, 1))
    ggsave(file.path(FIG_DIR, "car_chatgpt.png"), p_single, width = 10, height = 10, dpi = 300)
    ggsave(file.path(FIG_DIR, "car_chatgpt.pdf"), p_single, width = 10, height = 10)
    cat("  ✓ car_chatgpt\n")
  }
}

cat("\n✅ 事件研究完成！\n")
